command/agent/consul/check_watcher_test.go

```diff
@@ -194,6 +195,28 @@ func TestCheckWatcher_Healthy(t *testing.T) {
 	}
 }
 
+// TestCheckWatcher_Unhealthy asserts unhealthy tasks are restarted exactly once.
+func TestCheckWatcher_Unhealthy(t *testing.T) {
```

Comment:
I'm missing some context for this test - does it trigger the deadlock issue here?  Is it a relatively easy thing to test for?

Reply:
No, this test just asserts checks are only restarted once.

I added a new test for the deadlock in b4b2b42 and confirmed it does cause the deadlock before my changes.